src/ruby/puppetserver-lib/puppet/server/compiler.rb

```diff
@@ -14,11 +14,40 @@ def compile(request_data)
         node = create_node(processed_hash)
 
         catalog = Puppet::Parser::Compiler.compile(node, processed_hash['job_id'])
+
+        maybe_save(processed_hash, node.facts, catalog)
+
         catalog.to_data_hash
       end
 
       private
 
+      def maybe_save(processed_hash, facts, catalog)
+        nodename = processed_hash['certname']
+        persist = processed_hash['persistence']
+        options = processed_hash.
+                    slice("environment", "transaction_id").
+                    map {|key, val| [key.intern, val] }.to_h
+
+        if persist['facts']
+          if Puppet::Node::Facts.indirection.terminus_class.to_s == "puppetdb"
+            Puppet.override({trusted_information: processed_hash['trusted_facts']}) do
```

Comment:
This either needs to be `processed_hash['trusted_facts']['values']` or we need to pass in trusted facts separately, like we do with regular facts. I think the latter might be more correct, because it ensures that we save any trusted facts we pulled out, even if they weren't supplied in the request. (I don't think anyone wants to use the endpoint this way, but I want to cover our bases.)

Reply:
I played with this, and the way it is now, round-tripped trusted facts -- submitted in the request, saved to PDB, then queried on the next request -- come back wrapped in `values` when you pull them back out of PDB.

Reply:
👍